### PR TITLE
vdk-impala: Introduce checks for scd1 template

### PR DESCRIPTION
**Pull request by sbuldeev**
What:
Adding functionality to handle the scd1 template behavior if there are checks provided by the user

Why:
It is linked to https://github.com/vmware/versatile-data-kit/issues/1361

Tests: provided positive and negative regression tests

### CODE REVIEW

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:17]: I think we should allow users to omit passing check (and assume it's None). 

```suggestion
    check = job_arguments.get("check", None)
```

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:20]: Let's have a default as well. If not specified uses same as target_schema 
```suggestion
    staging_schema = job_arguments.get("staging_schema". target_schema)
```

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:25]: It's considered bad practice to use the base Exception class . 

It's better to use specific exception. For Illegal arguments ValueError is a frequently used

File: projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py:160]: You may want to add an assert that verifies that the target has not been changed.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:40]: Copying from staging table to target you probably do not want to reuse the same query. 

It works for this template because it's insert overwrite, but won't for the more complex templates. 

And we want to keep the code as for the next template we'd likely want to abstract teh common functionalities and reuse them.

**Review by tozka — COMMENTED** [review]
Looks good to me. I made some comments.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (murphp15) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:34]: This probably is not an issue with this PR but my understanding of VDK. 
Here the insert statement is created with target table and schema. 
However how does it know what the source table and scheme is?

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (murphp15) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:41]: If you are trying to insert into the target table should it not be target_schema and target_view instead of source_schema and source_view?

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:17]: Do we need passing a default value as None? The default behavior of dict.get("key") is to return None if the key is not present.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:20]: Wouldn't this cause an error? In the current logic the table name will be the same for both target and staging schemas.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py]: Changed that.

File: projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py:163]: Added the suggested assert.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:39]: I will have to think of a more generic way to transfer the data in order to fit all templates needs.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:33]: I expect all the arguments that are not passed explicitly to be inherited from the arguments passed to the template call.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:40]: On this step I want to insert from the staging table(where checks are performed) into target table. Therefore I only set the source to the staging table, because I want to avoid using the source view.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:16]: Oh, I forgot, In this case I guess no.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:39]: Add a TODO here if you are not ready to solve it now. It's easier to figure out what is "common" after a second instance of something.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:19]: That's true. I didn't realize. 

But it's better UX to require as few users as possible to get started with. This means as many as possible parameters to have defaults.
Wouldn't be better to prefix the table as well? If one uses a staging db for multiple purposes, this way it'd be easy to see those tables are generated for checks. Maybe we can add prefix "vdk_check_" ? 

What do you think ?

**Comment by tozka** [comment]
I don't know if you've noticed but there are some failing tests: https://gitlab.com/vmware-analytics/versatile-data-kit/-/jobs/3552673519 

Have you be able to run them locally ? In IDE ?

**Comment by sbuldeev** [comment]
> I don't know if you've noticed but there are some failing tests: https://gitlab.com/vmware-analytics/versatile-data-kit/-/jobs/3552673519
> 
> Have you be able to run them locally ? In IDE ?

These tests are failing, because at some point the execute_tempalate behavior is trying to dump all the arguments and since there is a function passed as an argument this is not possible.

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:40]: Okay lets proceed without that for now. 
We've discussed internally a way to introduce such generic functionality, but it will take time to be implemented, so we prefer to stick with what we have and introduce it on the next template modification.

File: projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py:141]: To avoid situaitons in which another test populates the data in a table. Let's use distinct names

```
source_view="vw_dim_org_check_negative"
target_table ="dw_dim_org_check_negative"

```

File: projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py:116]: To avoid situaitons in which another test populates the data in a table. Let's use distinct names

```
source_view="vw_dim_org_check_positive"
target_table ="dw_dim_org_check_positive"

```

File: projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py
Comment (tozka) [inline projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:13]: Let's add a comment which explains what the step does at high level (e.g what is its purpose). To help future maintainers.

Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/src/vdk/plugin/impala/templates/load/dimension/scd1/02-handle-quality-checks.py:17]: Added high level description.

File: projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py]: Done.

File: projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py
Comment (sbuldeev) [reply projects/vdk-plugins/vdk-impala/tests/functional/template_regression_test.py]: Done

**Review by tozka — APPROVED** [review]
Looks good to me. 
Thanks for the hard work on this change.

**Comment by tozka** [comment]
@sbuldeev, before merging do ping me so I can walk you through the proper process for merging a change. 

A) It's fairly important that there's one commit (in this case particularly) - so you need to use Squash option 
B) It's important that the commit description is the same as the PR description. 
   There also should be Signed-off-by: Humpty Dumpty <humpty.dumpty@example.com> in the commit description.

**Review by murphp15 — APPROVED** [review]
Looks good

**Pull request merged**
